### PR TITLE
[roku] Add power, powerState, player, and activeAppName

### PR DESCRIPTION
**Pull request by morph166955**
Adds a R/W channel for power, R/O channel for powerState for TV devices
Adds a R/O channel for the app name for all devices.
Adjusts minimum refresh timer to 1 second
Adds a player channel

### CODE REVIEW

**Comment by lsiepel** [comment]
Don’t forget to add thing upgrade instructions and regenerate  default i18n file

**Comment by morph166955** [comment]
> Don’t forget to add thing upgrade instructions and regenerate default i18n file

Sounds good. I need to go hunt down the thing upgrade instructions, haven't done them yet.

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (mlobstein) [inline bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:149]: If the TV is off you could return here to avoid all un-necessary queries below.
```suggestion
                            updateState(POWER, OnOffType.OFF);
                            return;
```

File: bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml
Comment (mlobstein) [inline bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml:58]: `power` could be moved to the top since it is the most important control for a TV.

**Review by mlobstein — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.roku/README.md
Comment (morph166955) [reply bundles/org.openhab.binding.roku/README.md:142]: My problem is with things like HabPanel and DSL rules.  I still get the app ID number.  Having a direct string channel resolves this issue.

File: bundles/org.openhab.binding.roku/README.md
Comment (morph166955) [reply bundles/org.openhab.binding.roku/README.md:30]: This is a typo on the readme.  I'm not lowering the default below 10.  I just opened up the minimum down to 1.  I have several rules that watch for power state, 10 seconds it just too long to wait.

File: bundles/org.openhab.binding.roku/README.md
Comment (morph166955) [reply bundles/org.openhab.binding.roku/README.md:132]: There is more than on/off coming from the device.  For example, when off my TVs say "Ready".  They can say different things depending on the low power mode settings.  It could be relevant to some, so I included it.

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (morph166955) [reply bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:353]: Unfortunately it is an older bug.  Some older devices don't accept PowerOn but they all accept POWERON.  Oddly enough, PowerOff works just fine.  There was a post somewhere about it but I can't find it right now.

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (morph166955) [reply bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:371]: I'll pull this out.  I had it in there before I added the on/off and didn't pull it out when I made the channel read-only.

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (morph166955) [reply bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:149]: I'll add it.

File: bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml
Comment (morph166955) [reply bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml:58]: Easy enough.

**Comment by morph166955** [comment]
I pushed a commit to cover most of the changes.  There are some comments above for discussion for the rest.

File: bundles/org.openhab.binding.roku/README.md
Comment (mlobstein) [reply bundles/org.openhab.binding.roku/README.md:142]: > My problem is with things like HabPanel, the API, and DSL rules.

Ah yes that can be an issue, I use this for DSL:
` var appName = Player_ActiveApp.stateDescription.options.filter[stateOption | stateOption.value == Player_ActiveApp.state.toString].get(0).getLabel()`

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (mlobstein) [reply bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:354]: If that is the case, the entry in roku.xml for buttonTv should be updated. A note in the README under the section `List of additional button commands for Roku TVs` would be nice too.

File: bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml
Comment (mlobstein) [inline bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml:150]: This can be set to readOnly
```suggestion
		<description>The Currently Running App on the Roku</description>
		<state readOnly="true"/>
```

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (mlobstein) [inline bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:146]: In case some devices report in all uppercase:
```suggestion
                        if (POWERON.equalsIgnoreCase(powerMode)) {
```

**Comment by morph166955** [comment]
Pushed updates from last review.

**Review by mlobstein — CHANGES_REQUESTED** [review]
Can you update this entry in roku.xml to `value="POWERON"`? Also update the corresponding line in i18n/roku.properties. Otherwise looks good.

image

**Comment by morph166955** [comment]
Done

**Review by mlobstein — APPROVED** [review]
LGTM, except still needing upgrade instructions.

**Comment by morph166955** [comment]
Circling back on this so we can get it closed up.  I haven't done thing update instructions yet, can anyone point me in the direction of what you're talking about so I can reference and finish up?

**Comment by lsiepel** [comment]
> Circling back on this so we can get it closed up. I haven't done thing update instructions yet, can anyone point me in the direction of what you're talking about so I can reference and finish up?

https://www.openhab.org/docs/developer/bindings/thing-xml.html#updating-thing-types

You can also look for some recent merged PR’s that add channels as they also have these instructions.

**Comment by morph166955** [comment]
Pushed an update.xml file.  Not sure how I can test it easily.

**Comment by lsiepel** [comment]
Also need to set the thingversion to the thing decleration. 

You can test to install the current version binding by the UI. Add the thing. Disable the binding. Put the updated binding jar in the addon folder. Now the previously added thing should have the new channels. 

So it would not be needed to re-create the thing

**Comment by morph166955** [comment]
Merged in the PR from @mlobstein to avoid any merge conflicts between the two PRs.

**Comment by morph166955** [comment]
Looks like it all compiled clean. I've been running the jar for weeks with no issues. I think this is good to merge unless anyone sees any issues.

**Comment by mlobstein** [comment]
> Looks like it all compiled clean. I've been running the jar for weeks with no issues. I think this is good to merge unless anyone sees any issues.

Agreeed, this is ready for final review. @jlaur, please review when you have time.

File: bundles/org.openhab.binding.roku/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.roku/README.md:53]: Can you provide some context here. What are the possible string values? I assume there are strings representing other values than on/off, since otherwise this channel would be fully redundant?

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/RokuBindingConstants.java
Comment (jlaur) [inline bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/RokuBindingConstants.java:86]:
```suggestion
    public static final String POWER_ON = "PowerOn";
```

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:151]: Could be considered:
```suggestion
                        updateState(POWER, OnOffType.from(POWERON.equalsIgnoreCase(powerMode)));
```

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:155]: It seems redundant to log exception message when you are also logging full exception with stack trace:
```suggestion
                        logger.debug("Unable to retrieve Roku device-info.", e);
```

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:353]:
```suggestion
        } else if (POWER.equals(channelUID.getId())) {
```

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (jlaur) [reply bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:358]: Perhaps `RokuBindingConstants.POWERON` could then be adapted to "POWERON" accordingly and used here as well?

File: bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml
Comment (jlaur) [inline bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml:159]:
```suggestion
		<description>The currently running app on the Roku</description>
```

File: bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml
Comment (jlaur) [inline bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml:235]:
```suggestion
		<description>Power state of the TV</description>
```

See also comment in README regarding this channel type. If this channel is really needed, perhaps some additional description would be helpful? Probably it should also be marked as advanced, since I would assume most users would be satisfied with the bi-directional `power` channel.

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks! I have added a few comments for consideration.

**Comment by morph166955** [comment]
Thank you for the review @jlaur. I'm out of town on vacation, I'll address them next week when I return. If anyone's feeling froggy and wants to submit the review changes to my branch to move this forward before I get home I can merge them in from my phone.

File: bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml
Comment (mlobstein) [inline bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml:13]: I did some additional testing... When adding a new thing, a duplicate channel error is seen due to the missing **thingTypeVersion**.

```suggestion
		<properties>
			<property name="thingTypeVersion">1</property>
		</properties>
```
See this for more info: https://www.openhab.org/docs/developer/bindings/thing-xml.html#updating-thing-types

File: bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml
Comment (mlobstein) [inline bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml:63]: Add thingTypeVersion to both thing types:
```suggestion
		<properties>
			<property name="thingTypeVersion">1</property>
		</properties>
```

**Review by mlobstein — CHANGES_REQUESTED** [review]
Additional changes are needed to complete the thing upgrade instructions.

File: bundles/org.openhab.binding.roku/README.md
Comment (morph166955) [reply bundles/org.openhab.binding.roku/README.md:53]: powerState provides more context when the device is in an idle state on some devices.  For example, my TV lists as "Ready" when in low power mode.  I've seen references to other states as well across the googleverse.  I included it for those who want specifics. I do agree the power channel will work for most, I'm just the type of person who will provide more info if I have it on something because I have no idea if it will help someone down the line.

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (morph166955) [reply bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:353]: This is probably something that needs to be changed across all of the channels under handleCommand.  I used the original to keep it in line with the rest of them.

File: bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml
Comment (morph166955) [reply bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml:159]: like handleCommand above, I kept this in line with the other channel-types and how they were cased.

**Comment by morph166955** [comment]
I think I got everything.  I had a stupid mistake when doing the push because I had to rebase to compile cleanly.  Please let me know if something got lost in the mix.

File: bundles/org.openhab.binding.roku/README.md
Comment (mlobstein) [reply bundles/org.openhab.binding.roku/README.md:53]: It might be good to include the possible values in parentheses:
```suggestion
| powerState         | String               | The current power state for the TV (ie PowerOn, DisplayOff, Ready, etc.).                                                                                                               |
```

File: bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml
Comment (mlobstein) [inline bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml:234]:
```suggestion
	<channel-type id="powerState" advanced="true">
```

**Review by mlobstein — APPROVED** [review]
LGTM

File: bundles/org.openhab.binding.roku/README.md
Comment (jlaur) [reply bundles/org.openhab.binding.roku/README.md:53]: Also "(ReadOnly)" is missing.

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (jlaur) [reply bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:348]: I agree it's out of scope to fix existing and you don't have to do that, but there is no reason to "keep it in line with potential `NullPointerException`" for new code.

File: bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml
Comment (jlaur) [inline bundles/org.openhab.binding.roku/src/main/resources/OH-INF/thing/roku.xml:238]: If you know at least some of the available states, you could add them as options here as this would both document them and make them available in the UI (and translateable):
```suggestion
		<state readOnly="true">
		    <options>
		        <option value="PowerOn">Power On</option>
		        <option value="DisplayOff">Display Off</option>
		    </options>
		</state>
```

etc.

**Comment by morph166955** [comment]
Pushed a commit with the final xml/readme changes.  I also reran the i18n script to make sure it's as current as possible.

**Comment by lolodomo** [comment]
@jlaur : have you time for a final review before milestone 2 is generated ?

**Comment by jlaur** [comment]
> have you time for a final review before milestone 2 is generated ?

Sure, I actually think I'm up to date with it. @morph166955 - can you check this comment which is still open: https://github.com/openhab/openhab-addons/pull/15542#discussion_r1314020487

**Comment by morph166955** [comment]
Is the request to just convert that to a constant at that point?

File: bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java
Comment (jlaur) [reply bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:353]: To be clear - question is if this constant:
```java
public static final String POWER_ON = "PowerOn";
```

could be changed to:
```java
public static final String POWER_ON = "POWERON";
```

and used here as well? Or if they are meant to be different?

Also there is @mlobstein's comment above to consider.

Comment (morph166955) [reply bundles/org.openhab.binding.roku/src/main/java/org/openhab/binding/roku/internal/handler/RokuHandler.java:353]: Tracking now. Yes, I believe they all should be changed to the "POWERON".  While the lower case works on most devices, an old bug would exclude some, which is obviously less than desirable.

**Comment by morph166955** [comment]
Commit pushed to fix the POWERON vs PowerOn comment.  I think this resolves everything.  Sorry for taking so long, I have been out of town for a while.

File: bundles/org.openhab.binding.roku/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.roku/README.md:90]:
```suggestion
POWERON _(NOTE: POWERON needs to be completely capitalized due to a bug with older Roku devices)_
```

If not parenthesis, consider some other separator as it looks a bit compacted:

image

**Comment by jlaur** [comment]
> Sorry for taking so long, I have been out of town for a while.

No problem at all. Welcome back. 🙂 I have added one last small comment for the README.

**Comment by morph166955** [comment]
README updated.

**Review by jlaur — APPROVED** [review]
LGTM

**Comment by morph166955** [comment]
Thank you!

**Pull request merged**
